### PR TITLE
PE-635 Safe EntityID

### PR DESCRIPTION
**Pull request by arielmelendez**
Suggested Reviewing Order:
• entity_id.ts: the class to enforce EntityID format and the EID() convenience function for instantiating EntityIDs.
• entity_id.test.ts: Unit tests
• types.ts: Observe the changes to the aliases. We migrated the existing EntityID, which was a Union Type of the existing entity ID classes, to AnyEntityID since I couldn't just use EntityID in places where a FolderID could be used (since the value might also be a placeholder constant for root folders). This may be something we want to improve upon, e.g. a different constant set for root folders' parent folder IDs, like all 0's.
• arfs_entities.ts: Observe that ArFSEntity, ArFSDriveEntity, and ArFSFileFolderEntity were hoisted out of ardrivecore-js in order to make it them use proper EntityID-aliased classes rather than the string alias that was in types.ts.
    • Also observe that sync_status has been removed
    • Notice that the ArFSDriveEntity has `rootFolderId: FolderID | ENCRYPTED_DATA_PLACEHOLDER_TYPE;` This is because of the constant choice that was made for the SAFE entity types. We may want to improve upon this in the future.

### CODE REVIEW

File: src/types/entity_id.ts
Comment (matibat) [inline src/types/entity_id.ts:19]: `this.entityId` is already a primitive string
```suggestion
		return this.entityId === `${entityId}`;
```

File: src/types/entity_id.test.ts
Comment (matibat) [inline src/types/entity_id.test.ts:65]: Maybe use `isEqualTo()` here

Comment (matibat) [reply src/types/entity_id.test.ts:65]: nevermind, that function is already being tested

**Review by matibat — COMMENTED** [review]
Half-reviewed it

File: src/types/entity_id.ts
Comment (matibat) [reply src/types/entity_id.ts:19]: This is also valid
```suggestion
		return `${this}` === `${entityId}`;
```

File: src/ardrive.ts
Comment (matibat) [inline src/ardrive.ts:240]: We cannot use `isEqualTo()` because parentFolderId is nullable, right?
```suggestion
		if (originalFileMetaData.parentFolderId.isEqualTo(newParentFolderId)) {
```

File: src/ardrive.ts
Comment (matibat) [inline src/ardrive.ts:298]: We cannot use `isEqualTo()` because parentFolderId is nullable, right?
```suggestion
		if (originalFileMetaData.parentFolderId.isEqualTo(newParentFolderId)) {
```

File: src/ardrive.ts
Comment (matibat) [inline src/ardrive.ts:348]:
```suggestion
		if (folderId.isEqualTo(newParentFolderId)) {
```

File: src/ardrive.ts
Comment (matibat) [inline src/ardrive.ts:363]:
```suggestion
		if (originalFolderMetaData.parentFolderId.isEqualTo(newParentFolderId)) {
```

File: src/ardrive.ts
Comment (matibat) [inline src/ardrive.ts:415]:
```suggestion
		if (folderId.isEqualTo(newParentFolderId)) {
```

File: src/ardrive.ts
Comment (matibat) [inline src/ardrive.ts:430]: Same here, parentFolderId is nullable (?)
```suggestion
		if (originalFolderMetaData.parentFolderId.isEqualTo(newParentFolderId)) {
```

File: src/utils/arfs_builders/arfs_drive_builders.ts
Comment (matibat) [inline src/utils/arfs_builders/arfs_drive_builders.ts:22]: Wy can't we? :thinking:

**Review by matibat — COMMENTED** [review]
Left a few suggestions, actually as questions

File: src/types/entity_id.ts
Comment (matibat) [inline src/types/entity_id.ts:1]: Are these always uncapitalized? Can we go case `i`nsentitive for hex?
```suggestion
const entityIdRegex = /^([a-f]|[0-9]){8}-([a-f]|[0-9]){4}-([a-f]|[0-9]){4}-([a-f]|[0-9]){4}-([a-f]|[0-9]){12}$/i;
```

Comment (matibat) [reply src/types/entity_id.ts:1]: We can poorly simplify this monster regexp, but I think it's worth it :)
```suggestion
const entityIdRegex = /^([a-f\d]){8}-([a-f\d]){4}-([a-f\d]){4}-([a-f\d]){4}-([a-f\d]){12}$/;
```

Comment (matibat) [reply src/types/entity_id.ts:1]: We could go even further, but it might lack redability
```suggestion
const entityIdRegex = /^[a-f\d]{8}-([a-f\d]{4}-){3}[a-f\d]{12}$/i;
```

**Comment by arielmelendez** [comment]
Looks like I've broken something:
```
yarn ardrive list-all-drives -w ./CLI_TEST_arweave-keyfile-g1hzNXVbh2M6LMQSUYp7HgkgxdadYqYEfw-HAajlms0.json
Invalid entity ID 'ENCRYPTED'!'
```

File: src/utils/arfs_builders/arfs_drive_builders.ts
Comment (arielmelendez) [reply src/utils/arfs_builders/arfs_drive_builders.ts]: a112923

File: src/types/entity_id.ts
Comment (arielmelendez) [reply src/types/entity_id.ts:1]: > Are these always uncapitalized? Can we go case `i`nsentitive for hex?

From the Wikipedia entry on UUID:
```
RFC 4122 Section 3 requires that the characters be generated in lower case, while being case-insensitive on input.
```
So I think we can be forgiving on case.

File: src/arfs_entities.ts
Comment (fedellen) [inline src/arfs_entities.ts:58]: Is this necessary here since we added `EncryptedEntityId`?

File: src/utils/arfs_builders/arfs_drive_builders.ts
Comment (fedellen) [inline src/utils/arfs_builders/arfs_drive_builders.ts:76]: Why is this check in the public drive builder but not the private? Do we need this?

File: src/types.ts
Comment (fedellen) [inline src/types.ts:15]: I don't think this is technically correct here. And this is also causing the type issues downstream where you have to cast things `as EntityID` in a couple places in the PR.

A folder's `EntityID` will never be `'root folder'`. That string is only ever used as a place holder for the `parentFolderId` field when it is missing. This case should only happen on the root folder of a drive. So, `folderId` will never be root folder, only the `parentFolderId` field.

```suggestion
export type FolderID = EntityID;
```

Comment (arielmelendez) [reply src/types.ts:15]: Yep those placeholder values were causing exceptions at runtime. I'm going to try the same pattern we used for EncryptedEntityId.

File: src/arfs_entities.ts
Comment (arielmelendez) [reply src/arfs_entities.ts:58]: 2738016f1886bf5b426e20a42adfbad322765fd6

File: src/utils/arfs_builders/arfs_drive_builders.ts
Comment (arielmelendez) [reply src/utils/arfs_builders/arfs_drive_builders.ts]: befad39

File: src/types.ts
Comment (arielmelendez) [reply src/types.ts]: 2738016f1886bf5b426e20a42adfbad322765fd6

File: src/types/entity_id.ts
Comment (arielmelendez) [reply src/types/entity_id.ts]: c90ae71

Good stuff @matibat !

File: src/ardrive.ts
Comment (matibat) [reply src/ardrive.ts:363]: parent folders are nullable

File: src/arfsdao.ts
Comment (matibat) [inline src/arfsdao.ts:745]: is this `toString`:
 - window.toString
 - this.toString
 - Object.prototype.toString
 - none of them (?
 
 let's maybe do the mapping with an arrow function

File: src/utils/arfs_builders/arfs_drive_builders.ts
Comment (matibat) [inline src/utils/arfs_builders/arfs_drive_builders.ts:217]: So any encrypted entity id will equal any other unrelated encrypted entity id? :thinking:

Comment (matibat) [reply src/utils/arfs_builders/arfs_drive_builders.ts:217]: Does that comparison happen somewhere?

File: src/utils/arfs_builders/arfs_folder_builders.ts
Comment (matibat) [inline src/utils/arfs_builders/arfs_folder_builders.ts:16]: Beware all root folder ids will equal for any drive

**Review by matibat — CHANGES_REQUESTED** [review]
It's looking great! I'm happy with the implementation of `toPrimitive`.

I left a pair of questions on which I see a possible risk, but maybe there isn't. I'll approve after those last comments are addressed :+1: 
_requesting changes because of the `map(toString)` which I think it actually affects that query_

File: src/ardrive.ts
Comment (matibat) [reply src/ardrive.ts]: exactly.

File: src/utils/arfs_builders/arfs_folder_builders.ts
Comment (arielmelendez) [reply src/utils/arfs_builders/arfs_folder_builders.ts:16]: We never query against this. It's just for display purposes. But yes it's sketchy and I don't love it.

File: src/utils/arfs_builders/arfs_drive_builders.ts
Comment (arielmelendez) [reply src/utils/arfs_builders/arfs_drive_builders.ts:217]: Similar to root folder this is just for display purposes for "fail safe decryption" of private entities for the list-all-drives command.

File: src/arfsdao.ts
Comment (arielmelendez) [reply src/arfsdao.ts]: Let's not leave it to chance 😬 2663b2b

File: src/utils/arfs_builders/arfs_folder_builders.ts
Comment (fedellen) [inline src/utils/arfs_builders/arfs_folder_builders.ts:12]: This would be a utility type that creates a placeholder for the parentFolderId field when its missing; which presumably means this folder is the root folder of the drive.

File: src/ardrive.ts
Comment (matibat) [reply src/ardrive.ts:348]: I'm fine with the current change

File: src/ardrive.ts
Comment (matibat) [reply src/ardrive.ts:415]: I'm fine with the current change

**Review by matibat — APPROVED** [review]
Thanks for your fixes! LGTM now!

File: src/utils/arfs_builders/arfs_folder_builders.ts
Comment (arielmelendez) [reply src/utils/arfs_builders/arfs_folder_builders.ts:12]: yea lemme update this

File: src/types/entity_id.ts
Comment (matibat) [reply src/types/entity_id.ts]: This is redundant, but it's not the end of the world

File: src/utils/arfs_builders/arfs_folder_builders.ts
Comment (arielmelendez) [reply src/utils/arfs_builders/arfs_folder_builders.ts]: 8bca2e9

**Review by fedellen — APPROVED** [review]

**Pull request merged**
